src/macros-by-example.adoc

```diff
@@ -0,0 +1,827 @@
+the right circumstances the density wins can be significant.
+
+
+=== Macro Shapes
+
+We can now introduce the final kind of input constraint, macro-shaped parameters.  To understand
+the motivation, consider modeling a scatter-plot as a list of points:
+
+----
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+Lists like these exhibit a lot of repetition. Since we already have a `point` macro, we can
+eliminate a fair amount:
+
+----
+[(:point 3 17), (:point 395 23), (:point 15 48), (:point 2023 5), …]
+----
+
+This eliminates all the ``x``s and ``y``s, but leaves repeated macro invocations.  We can try to
+wrap this in another macro, but we find the type constraints insufficient, since the tightest we
+can go is `*struct*`, and things aren’t really any better:
+
+[{mrk}]
+----
+(*public* scatterplot [(*struct\...* points)]
+  [points])
+----
+----
+(:scatterplot (:point 3 17) (:point 395 23) (:point 15 48) (:point 2023 5) …)
+----
+
+What we’d like is to build eliminate the `point` calls and just write a stream of pairs, something
+like:
+
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+----
+
+We can achieve exactly that with a macro-shaped parameter, in which we use the `point` macro as a
+pseudo-type:
+
+[{mrk}]
+----
+(*public* scatterplot [(*point\...* points)]
+  [points])
+----
+----
+(:scatterplot (3 17) (395 23) (15 48) (2023 5) …)
+=>
+[{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}, …]
+----
+
+What’s happening here is that each argument S-expression like `(3 17)` is _implicitly an
+E-expression_ invoking the `point` macro.  The argument mirrors the shape of the inner macro,
+without repeating its name.  Further, expansion of the implied ``point``s happens automatically,
+so the overall behavior is just like the preceding struct-based variant and the `points`
+parameter produces a stream of structs.
+
+As the text notation here suggests, macro-shaped parameters are tagless, and the binary form also
+elides any opcodes mentioning `point` and just encodes its arguments with minimal delimiting.
+
+Macro types can be combined with any cardinality, following the same rules as before.  Note that
+when combined with `***` or `*+*` this requires callers to use two layers of delimiting
+containers, and outer delimiter for the parameter-stream as a whole, and an inner delimiter for
+each macro instance:
+
+[{mrk}]
+----
+(*public* scatterplot
+  [(*point** points), (*string* x_label), (*string* y_label)]
+  { points: [points], x_label: x_label, y_label: y_label })
+----
+----
+(:scatterplot ((3 17) (395 23) (15 48) (2023 5)) "hour" "widgets")
+=>
+{
+  points: [{x:3, y:17}, {x:395, y:23}, {x:15, y:48}, {x:2023, y:5}],
+  x_label: "hour",
+  y_label: "widgets"
+}
+----
+
+TODO Describe binary representation
+
+CAUTION: Because these are tagless encodings, we cannot use macros in place of the delimiting
+containers, so unfortunately we cannot generate a point or series of points through a macro.
```

Comment:
Actually, I don't think I understand this advisory as well--I think what you're saying is that the only thing allowed are arguments that denote the macro invocation (specifically the s-expression containing the arguments.  One could not use anything else in its stead.  Is that right?

Reply:
OK I see the problem, my CAUTION confuses two things.   It's repeating something already covered, that you can't replace a delimiting-list with a macro invocation.  I'll make that more clear, but what it _should_ be saying is that you can't use a macro as an element of the delimiting-list, either (because the point of the macro-shape is to avoid the tags).

Let me add an examples here, hopefully it'll clarify.

Reply:
> It's repeating something already covered, that you can't replace a delimiting-list with a macro invocation

I might have already said it, but this seems like a good reason for delimiting containers and macros/macro-shapes to have different syntax.

> you can't use a macro as an element of the delimiting-list

Just curious—is this because we think it's a bad idea or because there's a technical limitation that prevents us from figuring out how a macro output should be converted to macro-shape arguments?

It seems like it would be convenient to be able to do something like this:
```
(*public* point
  [(*tagless int* x), (*tagless int* y)]
  {x: x, y: y})

(public scatterplot
  [(point* points), (string x_label), (string y_label)]
  { points: [points], x_label: x_label, y_label: y_label })

(public cartesian_product
  [(any+ a), (any+ b)]
  (for_each a (for_each b (a b)))
```
```
(:scatterplot [(1 1), (:cartesian_product [1, 2], [3, 4])] "hour" "widgets")
  ⇒ (:scatterplot [(1 1), [(1 3), (1 4), (2 3), (2 4)]] "hour" "widgets")
  ⇒ (:scatterplot [(1 1), (1 3), (1 4), (2 3), (2 4)] "hour" "widgets")
  ⇒ {
      points: [{x:1, y:1}, {x:1, y:3}, {x:1, y:4}, {x:2, y:3}, {x:2, y:4}],
      x_label: "hour",
      y_label: "widgets"
    }
```
I know that cartesian product is unlikely to be used in the real world, but the point is that being able to use the outputs of one macro as the inputs for a macro-shape parameter seems like it would make it a lot easier to compose macros.

Without this ability, you'd need to write a separate logic and add separate parameters in the `scatterplot` function to handle to the cartesian product logic. It's _kind of_ like the difference between these Rust functions: 
```rust
fn generateSpeech(str_utterances: Vec<&str>, string_utterances: Vec<String>, utf8bytes_utterances: Vec<&[u8]>) -> Mp3Bytes;
fn generateSpeech(utterances: Vec<impl Into<String>>) -> Mp3Bytes;
```